### PR TITLE
New project layout

### PR DESCRIPTION
**Pull request by chrisnojima**
This PR will finally move our project structure to be more 'vanilla' and remove cross symlinking the shared code.

tldr; desktop and react-native folders have been moved under shared (we can change this further later). We now share a single package.json

Original PR here: https://github.com/keybase/client/pull/4946

Split into 4 different sections:

- [x] https://github.com/keybase/client/pull/5236/commits/609a0d7a5406272613b53fb7aec963ec9b5e86c3 Most of the files are simple moves / renames. Some flowissues go away as flow was confused by the symlinks
- [x] https://github.com/keybase/client/pull/5236/commits/a1bdf7374e01672d929601b57ebac8729a5233a2 Updating Jenkins / babel / webpack. A lot of the webpack funkiness is now gone. Babel is driven by a BABEL_ENV that package.json handles which pulls in different presets for RN
- [x] https://github.com/keybase/client/pull/5236/commits/acc17502c0b190e8ab10be5c75fd0a3d347863ab The electron packaging script had to be touched a bunch to make it work with the different paths
- [x] https://github.com/keybase/client/pull/5236/commits/ddfd76b33dca01ef7dd718a4e2d76be0980839dc Fixes the common input components flexbox issues in RN. In the process of making this work I had to update RN and it now gives you a redbox when you have the flexbox issues instead of just rendering it in a funky way

### CODE REVIEW

File: shared/login/register/gpg-sign/index.render.native.js [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// @flow
+const empty = () => null
```
Comment (chrisnojima) [inline shared/login/register/gpg-sign/index.render.native.js]: Updated redux complains if you connect an empty component

File: Jenkinsfile [hunk 2]
```diff
@@ -122,15 +122,12 @@ helpers.rootLinuxNode(env, {
...
-                                }
                                 dir("shared") {
-                                    sh "../desktop/node_modules/.bin/flow"
-                                }
-                                sh "desktop/node_modules/.bin/eslint ."
-                                dir("desktop") {
-                                    sh "npm test"
+                                    sh "node --version"
+                                    sh "yarn install --pure-lockfile --verbose --prefer-offline --no-emoji --no-progress"
+                                    sh "yarn run flow"
```
Comment (chrisnojima) [inline Jenkinsfile]: new helpers for flow/lint/etc

File: Jenkinsfile [hunk 3]
```diff
@@ -122,15 +122,12 @@ helpers.rootLinuxNode(env, {
...
-                                dir("desktop") {
-                                    sh "yarn"
-                                }
                                 dir("shared") {
-                                    sh "../desktop/node_modules/.bin/flow"
-                                }
-                                sh "desktop/node_modules/.bin/eslint ."
-                                dir("desktop") {
-                                    sh "npm test"
+                                    sh "node --version"
```
Comment (chrisnojima) [inline Jenkinsfile]: Just so we can debug node issues

File: desktop/package.json [hunk 4]
```diff
@@ -1,133 +0,0 @@
-{
```
Comment (chrisnojima) [inline desktop/package.json]: This and the RN one are now merged into the shared/package.json

File: shared/.babelrc [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    "plugins": [
+        "babel-plugin-transform-runtime",
+        "react-hot-loader/babel",
+        ["babel-plugin-transform-builtin-extend", {
+            "globals": ["Error"]
+        }]
+    ],
+    "env": {
+        "react-native": {
+          "presets": ["react-native"],
```
Comment (chrisnojima) [inline shared/.babelrc]: We share a babelrc now so we can override it using an env var

File: shared/desktop/webpack.config.base.js [hunk 6]
```diff
@@ -49,15 +49,7 @@ module.exports = {
     libraryTarget: 'commonjs2',
   },
   resolve: {
-    modulesDirectories: [path.join(__dirname, 'node_modules')],
```
Comment (chrisnojima) [inline shared/desktop/webpack.config.base.js]: don't have to deal with this funkiness anymore

File: shared/desktop/yarn-helper.js [hunk 7]
```diff
@@ -177,64 +177,29 @@ const commands = {
 }
 
 function postInstall () {
-  if (process.platform === 'win32') {
```
Comment (chrisnojima) [inline shared/desktop/yarn-helper.js]: don't have to deal with the shared code being a symlink. renderer/fonts isn't being fixed up but i think that'd only affect devving on windows. we can revisit this

File: shared/desktop/yarn-helper.js [hunk 8]
```diff
@@ -177,64 +177,29 @@ const commands = {
...
-  } else {
-    // Making a shim module. TODO use rn's make-shim instead. This is to make msgpack requires work but allow the fallback to purepack (in node-framemsgpack code)
-    exec("mkdir -p node_modules/msgpack; echo 'module.exports = null' > node_modules/msgpack/index.js; echo '{\"main\": \"index.js\"}' > node_modules/msgpack/package.json")
-  }
-
-  const materialPath = 'node_modules/material-ui/package.json'
-  const materialUIJson = JSON.parse(fs.readFileSync(materialPath, 'utf8'))
-  materialUIJson.peerDependencies = {}
-  fs.writeFileSync(materialPath, JSON.stringify(materialUIJson, null, 4), 'utf8')
+  exec('node make-shim net')
```
Comment (chrisnojima) [inline shared/desktop/yarn-helper.js]: use the helper to make these shims

File: shared/desktop/package.js [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+import path from 'path'
+import webpack from 'webpack'
+import webpackConfig from './webpack.config.production.js'
+import {exec} from 'child_process'
+
+const filterAllowOnlyTypes = (...types) => ({
+  filter: f => types.some(type => f.endsWith(`.${type}`)),
+})
+
+const desktopPath = (...args) => path.join(__dirname, ...args)
```
Comment (chrisnojima) [inline shared/desktop/package.js]: helper to make this work relative to itself

File: shared/desktop/package.js [hunk 10]
```diff
@@ -0,0 +1,181 @@
+// @flow
```
Comment (chrisnojima) [inline shared/desktop/package.js]: this file is a move of the old one but cleaned up a little (wrapped in a main function and using some more es6 things)

File: shared/common-adapters/input.native.js [hunk 11]
```diff
@@ -148,7 +148,7 @@ class Input extends Component<void, Props, State> {
       fontSize: _headerTextStyle.fontSize,
       lineHeight: _lineHeight,
       backgroundColor: globalColors.transparent,
-      flex: 1,
+      flexGrow: 1,
```
Comment (chrisnojima) [inline shared/common-adapters/input.native.js]: with that breaking change in RN we likely have to do changes like this to fix the squished views. This updated version of RN actually gives you a redbox vs rendering something it knows is invalid

**Comment by chrisnojima** [comment]
Before this rolls out (if possible)

- [ ] Test electron prod builds on linux
- [ ] Test electron prod builds on windows
- [ ] Test electron prod builds on osx
- [ ] Testflight build of ios
- [ ] Test build of android

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is ready to look at. This review is large enough that I'd advocate for at least 2 thumbs thanks!

**Comment by chrisnojima** [comment]
Unclear why the tests are failing. Literally the same code as the other pr

**Comment by chrisnojima** [comment]
@keybase/react-hackers bump

File: shared/desktop/renderer/remote-component-loader.js [hunk 12]
```diff
@@ -177,7 +175,7 @@ const styles = {
     backgroundColor: globalColors.white,
   },
   loading: {
-    backgroundColor: globalColors.grey5,
+    backgroundColor: globalColors.grey,
```
Comment (chromakode) [inline shared/desktop/renderer/remote-component-loader.js]: :sheep: Unrelated visual tweak?

Comment (chrisnojima) [inline shared/desktop/renderer/remote-component-loader.js]: there never was a grey5! this is cause eslint / flow is working better everywhere

File: shared/.flowconfig [hunk 13]
```diff
@@ -1,33 +1,33 @@
...
 .*/node_modules/spdx-license-ids/.*
 .*/node_modules/ua-parser-js/.*
 .*/node_modules/y18n/.*
+.*/react/node_modules/.*
+.*/react/node_modules/.*
+../desktop/build/.*
+../desktop/release/.*
+../desktop/shared/.*
+../react-native/shared/.*
+./desktop/build/.*
```
Comment (chrisnojima) [inline shared/.flowconfig]: yup!

File: shared/.flowconfig [hunk 14]
```diff
@@ -1,33 +1,33 @@
...
+.*/node_modules/react/.*
 .*/node_modules/resolve/.*
+.*/node_modules/resolve/lib/.*
 .*/node_modules/sha.js/test/.*
 .*/node_modules/spdx-exceptions/.*
 .*/node_modules/spdx-license-ids/.*
 .*/node_modules/ua-parser-js/.*
 .*/node_modules/y18n/.*
+.*/react/node_modules/.*
+.*/react/node_modules/.*
```
Comment (chrisnojima) [inline shared/.flowconfig]: yup

**Comment by chrisnojima** [comment]
visdiff works (based on testing a branch off of a branch using this PR: https://github.com/keybase/client/pull/5267)

**Comment by chrisnojima** [comment]
@cjb please TAL

**Comment by cjb** [comment]
@chrisnojima LGTM
